DOC: Improve cross-links in thread safety documentation

Carreau opened the pull request:
Those are mostly cosmetic changes and extra cross-links to improve the documentation. In particular I'd like to use this page as an example of good top-level documentation for threading/freethreading in the py-threading-guide https://py-free-threading.github.io/

This also does a number of updates to the misc.rst file which was basically too indented, and missing a bunch of links. I think many of the info misc.rst on how to interop with C/C++ could be removed, and likely all the information on seterr/geterr moved to routines.err.rst but I don't want to include this moves as part of this PR as it's likely more subject to discussion.

There are minor updates in conf.py to avoid 1 sphinx warnings, and I know the thread safety backlink in geterr/seterr/...etc, are a bit overkill, but as numpy will be used as an example and in general is referred to as one of the core scientific Python project I thinks it's OK go go a bit further than usual.

[skip actions]
[skip cirrus]

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

Carreau commented on doc/source/user/misc.rst:
```diff
@@ -56,159 +68,178 @@ and ``'ignore'`` for ``underflow``.  But this can be changed, and it can be
 
 Interfacing to C
 ----------------
 Only a survey of the choices. Little detail on how each works.
```
I think most of what is below here could/should use some pruning. But that might be another PR.

Carreau commented:
Maybe tag as free-threading ?

rgommers commented:
sure, done

mattip reviewed (commented):
Thanks for thinking about the documentation. In general cross-linking to the same term should be limited to a single reference per page, since it is visually disturbing.

> In particular I'd like to use this page as an example of good top-level documentation

Cool!

> This also does a number of updates to the misc.rst file

Hah, surprised you found it at all. That page is really an orphan that probably should have been removed at some stage. I would prefer to not add to it, rather to remove from it and move the content elsewhere.

ngoldbaum commented on doc/source/reference/thread_safety.rst:
```diff
@@ -22,15 +22,30 @@ are not reproducible, let alone correct. It is also possible to crash the Python
-not release the GIL.
+not release the :term:`python:GIL`.
+
+Thread-local state
```
errstate is stored in a contextvar, so it's actually context-local state. It's async-safe and thread-safe.

Maybe you could call it `Context-local state`? I'm not sure if there's a term for it that asyncio or CPython more generally uses.

Also the printoptions are a contextvar as well.

ngoldbaum replied:
The custom allocator object is also context-local: https://numpy.org/devdocs/reference/c-api/data_memory.html#configurable-memory-routines-in-numpy-nep-49

Carreau reviewed (commented):
Some suggestions, in particular I'd like  your input as to wether put in `See Also`, or in `Note` the ref from the functions docstrings, or just not put it at all. 

I'm not a huge fan of See Also, as there is no way reading it to know it's about concurency; and same with saying "Concurency Note", if you search for "Thread" it does not match, so one may miss the text.

Carreau committed:
DOC: Improve cross-links in thread safety documentation

Those are mostly cosmetic changes and extra cross-links to improve the
documentation. In particular I'd like to use this page as an example of
good top-level documentation for threading/freethreading in the
py-threading-guide https://py-free-threading.github.io/

This also does a number of updates to the misc.rst file which was
basically too indented, and missing a bunch of links. I think many of
the info misc.rst on how to interop with C/C++ could be removed, and likely
all the information on seterr/geterr moved to routines.err.rst but I don't
want to include this moves as part of this PR as it's likely more
subject to discussion.

There are minor updates in conf.py to avoid 1 sphinx warnings, and I
know the thread safety backlink in geterr/seterr/...etc, are a bit overkill,
but as numpy will be used as an example and in general is referred to as
one of the core scientific Python project I thinks it's OK go go a bit
further than usual.

[skip actions]
[skip cirrus]

Carreau commented:
Squashed & rebased, sorry for the loss of history bur fixing each commit conflict with #30394 was annoying.

Carreau commented on doc/source/reference/routines.err.rst:
```diff
@@ -1,8 +1,86 @@
+
+::
+
+   >>> oldsettings = np.seterr(all='warn')
```
All exmamples re-written using errsate.

Carreau committed:
typo

ngoldbaum reviewed (approved):
Awesome, huge improvement!

@charris, I'm adding the backport-candidate label because I'd like to see these docs included in the 2.4.0 release. Feel free to disregard if you'd prefer not to.
